doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1747,69 +1718,67 @@ name of the variable (here :g:`n`) is chosen based on :g:`T`.
+   *from right to left* in the goal.  (Use one of
    the variants below to select an existential variable in a
    hypothesis.)  Counting starts at 1 and multiple occurrences of the
    same existential variable are counted multiple times.  Because this
    variant is not robust to slight changes in the goal, its use is
    strongly discouraged.
 
-.. tacv:: instantiate ( @natural := @term ) in @ident
-          instantiate ( @natural := @term ) in ( value of @ident )
-          instantiate ( @natural := @term ) in ( type of @ident )
+   :n:`in @ident`
+     Selects the hypothesis :n:`@ident`.
 
-   These allow to refer respectively to existential variables occurring in a
-   hypothesis or in the body or the type of a local definition (named :n:`@ident`).
+   :n:`in %|- *`
+     Selects all hypotheses and the goal.
 
-.. tacv:: instantiate
+   .. todo PR above is correct (in |- *)?
 
-   This tactic behaves functionally as :tacn:`idtac`.
+   :n:`in ( type of @ident )`
+    Selects existential variables with the type :n:`@ident` in a
+    :term:`local definition <context-local definition>`
 
-   .. deprecated:: 8.16
+   :n:`in ( value of @ident )`
+    Selects existential variables that have type :n:`@ident` in the body of
+    a :term:`local definition <context-local definition>`.
 
 .. tacn:: admit
-   :name: admit
 
    This tactic allows temporarily skipping a subgoal so as to
    progress further in the rest of the proof. A proof containing admitted
    goals cannot be closed with :cmd:`Qed` but only with :cmd:`Admitted`.
 
-.. tacv:: give_up
+   :tacn:`give_up` is a synonym for `admit`.
 
-   Synonym of :tacn:`admit`.
+.. todo PR: is give_up worth keeping?  See proof-mode.rst
 
-.. tacn:: absurd @term
-   :name: absurd
+.. tacn:: absurd @one_term
 
-   This tactic applies to any goal. The argument term is any proposition
+   The argument term is any proposition
    :g:`P` of type :g:`Prop`. This tactic applies False elimination, that is it
    deduces the current goal from False, and generates as subgoals :g:`∼P` and
    :g:`P`. It is very useful in proofs by cases, where some cases are
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction, such as
+   a hypothesis that is equivalent to `False` (from which any conclusion
+   can be proven).  This could also be (after all intros) a hypothesis that is
+   the negation of a singleton inductive type (e.g. :g:`True` or :g:`x=x`),
```

Comment:
The "negation of a singleton inductive type" case does not correspond to the behavior I observe when testing. And something that the current description omits seems to be that if the provided argument is the proof of a negation, then the tactic replaces the current goal by a goal asking to prove the non-negated statement.

Example:

```coq
Goal ~True -> forall A : Prop, A.
intro H.
contradiction H. (* Goal: True *)
```

Reply:
> The "negation of a singleton inductive type" case does not correspond to the behavior I observe when testing.

Can you be more specific?  That doesn't work or it should say something stronger?

I really know very little about the tactic, perhaps you could provide text or bullet points (and an example or two, if appropriate)?  It would probably take considerable experimentation and guesswork on my part to figure it out.

Reply:
> That doesn't work or it should say something stronger?

I would say that it works, but it is just incorrectly documented (it has always worked like this, at least since 8.6).

> I really know very little about the tactic, perhaps you could provide text or bullet points (and an example or two, if appropriate)? It would probably take considerable experimentation and guesswork on my part to figure it out.

Sure, I can do that.

- `contradiction` without argument looks in the local context for either `False` or (most common case) two hypotheses, one which is the negation of the other.
- `contradiction` with an argument will solve the goal if it is a proof of `False` and ask for a proof of the non-negated statement if it is a proof of a negated statement but fail with "not a contradiction" if it is neither (I'm saying this based on my examples).
- We could review the various uses of `contradiction` in the stdlib, but a first thing to note is that the form with an argument is quite unusual.

I'll add examples later on.